iterator.go

```diff
@@ -0,0 +1,61 @@
+package cfdns
+
+import (
+	"context"
+	"errors"
+)
+
+var Done = errors.New("done")
+
+// Iterator implements an iterator algorithm from a function that fetches
+// blocks of data. This allows having fixed-memory usage when reading
+// arbitrary-sized structs without leaking implementation details about
+// how to paginate consecutive blocks of data.
+type Iterator[T any] struct {
+	fetchNext FetchFn[T]
+	elements  []T
+	isLast    bool
+}
+
+type FetchFn[T any] func(ctx context.Context) (batch []T, last bool, _ error)
```

Comment:
optional:

Would it be better to make it private?
```suggestion
type fetchFn[T any] func(ctx context.Context) (batch []T, last bool, _ error)
```